kbs/docs/initdata.md

```diff
@@ -0,0 +1,237 @@
+Many TEE platforms supports TEE initdata field. The TEE initdata field could be set by
+the untrusted host when launching the TEE, and the field will be included in the
+TEE evidence for remote attestation.
+
+Platforms and corresponding field of the evidence
+- Intel TDX: `mr_config_id`, 48 bytes. Actually `mr_owner` and `mr_owner_config` have similiar
+attributes, but we select only `mr_config_id` for such use.
+- AMD SNP: `hostdata`, 32 bytes.
+- Arm CCA: `CCA_REALM_PERSONALIZATION_VALUE`, 64 bytes.
+- Intel SGX: `CONFIGID`, 64 bytes.
+
+When users want to deploy a TEE, they need to prepare an initdata metadata. The host
+(probably untrusted) SHOULD start TEE instance with initdata digest as TEE initdata.
+
+The software outside the TEE would deliver the initdata into the TEE in
+some way due to concrete architecture design. This spec does not define the concrete
+way to deliver initdata, but strongly recommend that before using the received
+initdata, software stack inside the TEE **MUST** check whether the digest (calculate
+again) of the received initdata is aligned with the one inside evidence.
+
+Other platforms, such as (v)TPM based platforms, can record the initdata digest
+by extending the PCR before using it. This way will also accomplish the integrity binding
+to the TEE evidence.
+
+### Data Canonicalization Algorithm
+
+When we use JSON or Toml to represent an initdata, control characters like
+line breaks that do not affect the semantics but will cause the original initdata
+data content to change. This fact will make it difficult for us to calculate a
+determined digest from different forms of a same initdata data. We need a canonicalization
+algorithm. With this algorithm, all different forms of a same initdata will be calculated
+into a same initdata digest. The difference of forms at least include the following
+
+1. The encoding of the initdata, e.g. JSON and TOML.
+2. Initdata metadata with different expressions but the same semantics, e.g. multiple line
+JSON vs one line JSON like the following
+```json
+{
+  "algorithm": "sha384",
+  "version": "0.1.0",
+  "data": {
+    "key1": "value1",
+    "key2": "value2"
+  }
+}
+```
+and
+```json
+{"algorithm":"sha384","version":"0.1.0","data":{"key1":"value1","key2":"value2"}}
+```
+
+The algorithm has 2 steps as following
+
+1. Transform the original initdata data, s.t. `data` field into JSON. This step requires converting
+different forms of initdata metadata into JSON without changing the semantics. This is practical
+because JSON, TOML and YAML all support key-value, string, and nested formats.
+2. Canonicalize the JSON resulted in step 1. The canonicalization follows [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785).
+This includes removing extra whitespace characters and ordering of object members.
+
+After performing the algorithm, the result string can be used to apply hash `algorithm`.
```

Comment:
Can we specify how the hashing occurs in detail?

I imagine it's something like:

```
x = hash.new(alg)
for (_, payload) in sorted_initdata_dict:
  x.update(payload)
digest = x.digest()
```

is that correct?

Reply:
Yes! As hash algorithm is for bytes stream, meaning that `update(whole-data)` and `update(one-piece) X n times` will get the same result. So I do not describe that.

Let me add this now

Reply:
Ah sorry. Ignore my previous comment. Here the thing `"After performing the algorithm"` would look like
```json
{"key1": "value1","key2": "value2"}
```

It is already a string. So we do not need iter upon the key-values again. The core idea of canonicalization is to finalize a string that is to get a semantically unambiguous dict in string form

Reply:
but wouldn't we get a unique hash if we iterate over a sorted dictionary? in this case we would not need a canonicalization in the spec.

it would unroll to 

```python
m = hashlib.sha256()
m.update(b"value1")
m.update(b"value2")
d = m.digest()
```

Reply:
Besides the values, we should also cover the keys.

Also, if we use the algorithm, the following two data will get the same hash which is not expected.
```json
{
  "key1":"value1",
  "key2":"value2"
}
```

```json
{
  "key1":"value1val",
  "key2":"ue2"
}
```

Reply:
ah I understand. would hash extension be an option?

```
>>> m = hashlib.sha256()
>>> m.update(b'value1')
>>> m.update(m.digest() + b'value2')
>>> m.hexdigest()
'5b848becb4a8d7b1515dbd43472cd3d66e7d027d83fa004d1bb158cf1a248802'
>>> m = hashlib.sha256()
>>> m.update(b'value1val')
>>> m.update(m.digest() + b'ue2')
>>> m.hexdigest()
'c2fb21bbe4093099b053b169c292a14956b7177db07544fb024910c5afdd5223'
```

But yes, I assumed the keys wouldn't be part of the hashing. ofc keys could also be supported in such an extension scheme, and we wouldn't require json transformation/normalization.

In any case, we probably want to specify the hashing algorithm explicitly (in pseudo code) in this spec. since client-side tooling would need to implement it.

Reply:
Yes. Hash extension could also be an option without involving formatting change.

So we just need to express in the pseudo-code
```
hasher = hash.new(alg)
state = b""
for key, value in sorted-data-dict:
  hasher.update(state)
  haser.update(key)
  state = hasher.digest()
  hasher.update(state)
  haser.update(value)
  state = hasher.digest()
```

This sounds better than we transform the original format if we want the date is one-level string-to-string map as the hash extension somehow works the same as `=` and `,` in a JSON, although it requires more hash calculations.

On the other hand, if we want to support nested map in future, we should also change the algorithm here as this algorithm might not support to express nested structure.

Up to now it works fine. I am just a little afraid it could have some influence when we change the `data` shape in future, but it is not enough to prevent me from thinking that extend hash is better for now.

wdyt?

Reply:
let's try to get some more perspectives from others. I don't have strong opinions on this, I'm just exploring options how to avoid a json transformation/normalization step and make no assumptions about the payload of each initdata entry. 

If we want to make initdata a top-level config entity with rich nested data structures, instead of a flat k/v map, then yes, I think you need to normalize it somehow.

Reply:
Hmm. From an implementation standpoint I might prefer leaning on an existing canonicalization scheme (especially if there are good libraries implementing it) rather than coming up with our plan.

Reply:
Higher level question about canonicalization: Why do we need it? When we check the binding of the initdata and the hw evidence, we expect to have the initdata.

I don't think we'll ever need to assemble the initdata from scratch in a way that matches the hash in the report. So is the issue really more about making sure the format of the initdata is preserved when it is transmitted from host->guest->trustee? Maybe there is a way to ensure that without defining canonicalization as clearly.

Reply:
@fitzthum Yes. I think the core issue is to preserve when transmitted from host->guest->trustee. Users might add some comments or line breaks to the initdata TOML they write for better maintainability. However, these contents themselves do not affect the semantics of the TOML. Based on this, canonicalization is proposed to keep their hash values ​​consistent during the transmission process.

Of course we can use a more strict way, means that the text should always be the same in format and contents byte-by-byte.

The two ways are another trade-off.


BTW the origin of canonicalization. The historical reason is when we tried to design initdata, we wanted to add initdata as a new [yaml object](https://github.com/kata-containers/kata-containers/blob/a67382fd00eab6ef3df7d6432352dbe6d2714047/docs/how-to/how-to-set-sandbox-config-kata.md#runtime-options) so we need canonicalization to ensure the key order of the serialized string. Regrettably kata config does not support obejct but only string and number, so the idea of yaml object is deprecated, while the canonicalization is still in hand.

Reply:
@fitzthum Yes. I think the core issue is to preserve when transmitted from host->guest->trustee. Users might add some comments or line breaks to the initdata TOML they write for better maintainability. However, these contents themselves do not affect the semantics of the TOML. Based on this, canonicalization is proposed to keep their hash values ​​consistent during the transmission process.

Of course we can use a more strict way, means that the initdata the whole should always be the same in format and contents byte-by-byte, which means it should be treated as a string.

The two ways are another trade-off.


BTW the origin of canonicalization. The historical reason is when we tried to design initdata, we wanted to add initdata as a new [yaml object](https://github.com/kata-containers/kata-containers/blob/a67382fd00eab6ef3df7d6432352dbe6d2714047/docs/how-to/how-to-set-sandbox-config-kata.md#runtime-options) so we need canonicalization to ensure the key order of the serialized string. Regrettably kata config does not support obejct but only string and number, so the idea of yaml object is deprecated, while the canonicalization is still in hand.

Reply:
Let me summrize two ways. 
1. Use canonicalization. Flexible for inline comments.
2. Do not  use canonicalization. Original plaintext of initdata data should be byte-by-byte same.

Please share your vote upon the two ways.

Reply:
I have no preference here.

Reply:
@Xynnn007 I vote for "Original plaintext of initdata data should be byte-by-byte same." - to keep the complexity and opportunities for bugs to a minimum.

If we'll ever run out of more important things to do, we could add canonicalization later.

Reply:
Ok. So the preference is 1 :0. Let me make it clear the algorithm to calculate an initdata digest by examples.

1. TOML style.
```toml
algorithm = "sha384"
version = "0.1.0"

[data]
"attestation-agent.json" = '''
{
"kbs_addr": "http://172.18.0.1:8080"
}
'''

"policy.rego" = '''
package agent_policy
import future.keywords.in
import future.keywords.every
import input
# Default values, returned by OPA when rules cannot be evaluated to true.
default CopyFileRequest := false
'''
```

digest = hash(
```
"attestation-agent.json" = '''
{
"kbs_addr": "http://172.18.0.1:8080"
}
'''

"policy.rego" = '''
package agent_policy
import future.keywords.in
import future.keywords.every
import input
# Default values, returned by OPA when rules cannot be evaluated to true.
default CopyFileRequest := false
'''
```
)

2. JSON style
```json
{
  "algorithm": "sha384",
  "version": "0.1.0",
  "data": {
    "attestation-agent.json": "{\"aa_kbc_params\": \"cc_kbc::http://127.0.0.1:8080\"}",
    "policy.rego": "package agent_policy\nimport future.keywords.in\nimport future.keywords.every\nimport input\n\n# Default values, returned by OPA when rules cannot be evaluated to true.\ndefault CopyFileRequest := false\ndefault CreateContainerRequest := false\ndefault CreateSandboxRequest := true\ndefault DestroySandboxRequest := true\ndefault ExecProcessRequest := false\ndefault GetOOMEventRequest := true\ndefault GuestDetailsRequest := true\ndefault OnlineCPUMemRequest := true\ndefault PullImageRequest := true\ndefault ReadStreamRequest := false\ndefault RemoveContainerRequest := true\ndefault RemoveStaleVirtiofsShareMountsRequest := true\ndefault SignalProcessRequest := true\ndefault StartContainerRequest := true\ndefault StatsContainerRequest := true\ndefault TtyWinResizeRequest := true\ndefault UpdateEphemeralMountsRequest := true\ndefault UpdateInterfaceRequest := true\ndefault UpdateRoutesRequest := true\ndefault WaitProcessRequest := true\ndefault WriteStreamRequest := false"
  }
}
```

digest = hash(
```
{
    "attestation-agent.json": "{\"aa_kbc_params\": \"cc_kbc::http://127.0.0.1:8080\"}",
    "policy.rego": "package agent_policy\nimport future.keywords.in\nimport future.keywords.every\nimport input\n\n# Default values, returned by OPA when rules cannot be evaluated to true.\ndefault CopyFileRequest := false\ndefault CreateContainerRequest := false\ndefault CreateSandboxRequest := true\ndefault DestroySandboxRequest := true\ndefault ExecProcessRequest := false\ndefault GetOOMEventRequest := true\ndefault GuestDetailsRequest := true\ndefault OnlineCPUMemRequest := true\ndefault PullImageRequest := true\ndefault ReadStreamRequest := false\ndefault RemoveContainerRequest := true\ndefault RemoveStaleVirtiofsShareMountsRequest := true\ndefault SignalProcessRequest := true\ndefault StartContainerRequest := true\ndefault StatsContainerRequest := true\ndefault TtyWinResizeRequest := true\ndefault UpdateEphemeralMountsRequest := true\ndefault UpdateInterfaceRequest := true\ndefault UpdateRoutesRequest := true\ndefault WaitProcessRequest := true\ndefault WriteStreamRequest := false"
  }
```
)


When writing the examples, I realize that there is another problem if we do not have canonicalization. Take JSON for example. If we do not have canonicalization, it is hard to say how many spaces before the first `{`.

Thus I am worrying about that we did not get a consensus upon "what should be included to calculate the digest". I once wanted only the initdata `data` field. If we want to use "byte-to-byte" scheme, a better way is to use the whole initdata (including metadata) to calculate the digest.

Before I re-write the related parts, please ensure we vote for plan
- "byte-to-byte" way. And the input to calculate hash digest is the whole initdata, rather than only initdata `data` field.

Reply:
I would also vote for skipping canonicalization for now. It's the simpler scheme that reduces cognitive overhead, since it's obvious what is being hashed.

I did not fully understand yet, why we would need to hash the full init-data body instead of the data payloads in a byte-to-byte scheme, can you explain?

to me it looks like the metadata contains specific instructions for a hashing/verifying function that is consuming various to-be-hashed elements.

Reply:
@mkulke Ok. It goes back to the old problem we've discussed in https://github.com/confidential-containers/trustee/pull/348#discussion_r1519304111. Let me take a simpler example. If the initdata body looks like
```json
{
  "algorithm": "sha384",
  "version": "0.1.0",
  "data": {
    "key": "value"
  }
}
```
or
```toml
algorithm = "sha384"
version= "0.1.0"

[data]
"key" = "value"
```

What are the exactly two inputs to calculate the digests? If we do not use the whole body as input, we should focus on `"data"`. Is the first `{\n\t"key": "value"\n\t}` (The spaces and `\n` here are annoying), or use [hash extension](https://github.com/confidential-containers/trustee/pull/348#discussion_r1519342579) upon every key-value pairs?

Reply:
ah, thank you for clarifying. I understand now. I initially assumed we could just hash the value part so a pseudocode algorithm would look like this:

```python
for _key, value in data.sorted_by(key):
   digest.extend_with(value)
```

`key` would be part of the metadata instructing the handler where this file should be instrumented to and not hashed. However, if we want to keep this information as part of a hash this doesn't work. we probably want to do this, an initdata object that instruments e.g. a policy file to wrong location should not produce a valid hash.

If we want to retain the key in the hash we need some form of layout schema indeed (kv pairs sorted by key?).

So, there could be either:

```python
for _key, value in data.sorted_by(key):
   digest.extend_with(key)
   digest.extend_with(value)
```

or hashing of the full body.

hashing the full body is the simpler option, I guess

Reply:
Yes. The only shortage is a same initdata body in different encoding (toml, json) would be different. But this seems fine now as no one is intended to use in this way. In future, if any scenario requires this, we can make up to the spec of the hash calculating part by updating version number as @danmihai1 suggests.